core/src/banking_stage.rs

```diff
@@ -177,81 +253,84 @@ impl BankingStage {
                     &msgs,
-                    unprocessed_indexes.to_owned(),
-                    transaction_status_sender.clone(),
-                    gossip_vote_sender,
+                    &old_unprocessed_indexes,
+                    my_pubkey,
+                    *next_leader,
                 );
-
-            new_tx_count += processed;
-
-            // Collect any unprocessed transactions in this batch for forwarding
-            rebuffered_packets += new_unprocessed_indexes.len();
-            Self::push_unprocessed(
-                &mut unprocessed_packets,
-                msgs,
-                new_unprocessed_indexes,
-                &mut dropped_batches_count,
-                batch_limit,
-            );
-
-            if processed < verified_txs_len {
-                let next_leader = poh_recorder.lock().unwrap().next_slot_leader();
-                // Walk thru rest of the transactions and filter out the invalid (e.g. too old) ones
-                #[allow(clippy::while_let_on_iterator)]
-                while let Some((msgs, unprocessed_indexes)) = buffered_packets_iter.next() {
-                    let unprocessed_indexes = Self::filter_unprocessed_packets(
-                        &bank,
-                        &msgs,
-                        &unprocessed_indexes,
-                        my_pubkey,
-                        next_leader,
-                    );
-                    Self::push_unprocessed(
-                        &mut unprocessed_packets,
-                        msgs,
-                        unprocessed_indexes,
-                        &mut dropped_batches_count,
-                        batch_limit,
-                    );
+                Self::update_buffered_packets_with_new_unprocessed(
+                    old_unprocessed_indexes,
+                    new_unprocessed_indexes,
+                )
+            } else {
+                let bank = poh_recorder.lock().unwrap().bank();
+                if let Some(bank) = bank {
+                    let (processed, verified_txs_len, new_unprocessed_indexes) =
+                        Self::process_received_packets(
+                            &bank,
+                            &poh_recorder,
+                            &msgs,
+                            old_unprocessed_indexes.to_owned(),
+                            transaction_status_sender.clone(),
+                            gossip_vote_sender,
+                        );
+                    new_tx_count += processed;
+                    // Out of the buffered packets just retried, collect any still unprocessed
+                    // transactions in this batch for forwarding
+                    rebuffered_packets_len += new_unprocessed_indexes.len();
+                    if processed < verified_txs_len {
+                        reached_end_of_slot =
+                            Some((poh_recorder.lock().unwrap().next_slot_leader(), bank));
+                    }
```

Comment:
how about moving this `if` to reside immediately after calling `process_received_packets()`? 

This code functions like proactive looping control code for the next iterations, so I rather don't want to put this in middle of bookkeeping of _this_ iteration if not required:

```rust
let (processed, verified_.., ...) = Self::process_...();
if processed < verified_... {
    // mark subsequent iterations as being skipped
    reached_... = ...
}

new_tx_count += processed;
// Out of the buffered packets just retried, collect any still unprocessed
// transactions in this batch for forwarding
rebuffered_packets_len += new_unprocessed_indexes.len()
let has_more_ ...

if let Some(test_fn) = ... {
    ...
}

has_more_....
```

(also note that my delicate preference is encoded in as a form of chunking by blank lines... lol)

Because of the current location of this code, I wondered while `return true/false` was needed inside this `if`

also, it seems that data dependency allows this. :)